ambry-store/src/test/java/com.github.ambry.store/IndexTest.java

```diff
@@ -483,22 +488,178 @@ public void hardDeleteTest() throws InterruptedException, IOException, StoreExce
+    idsToDelete.clear();
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : idsToDelete) {
+      addDeleteEntry(id);
+    }
+    verifyEntriesForHardDeletes(idsToDelete, false);
+    reloadIndex(false);
+    assertTrue("Hard deletes should have been paused ", index.hardDeleter.isPaused());
+    assertEquals("Hard delete should not have progressed ", expectedProgress, index.hardDeleter.getProgress());
+
+    Set<MockId> newIdsToDelete = new HashSet<>();
+    addPutEntries(3, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    newIdsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    addPutEntries(3, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    newIdsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : newIdsToDelete) {
+      addDeleteEntry(id);
+    }
+    idsToDelete.addAll(newIdsToDelete);
+    assertTrue("Hard deletes should have been paused ", index.hardDeleter.isPaused());
+    verifyHardDeleterThreadState(Thread.State.WAITING, hardDeleterSleepTimeWhenCaughtUpMs + 1,
+        hardDeleterSleepTimeWhenCaughtUpMs / 10);
     assertEquals("Hard delete has not processed all keys", expectedProgress, index.hardDeleter.getProgress());
+    verifyEntriesForHardDeletes(idsToDelete, false);
+    // resume and verify new entries(both before and after reload) have been hard deleted
+    index.hardDeleter.resume();
+    advanceTime(SystemTime.getInstance().milliseconds() + 2 * Time.MsPerSec * Time.SecsPerDay);
+    expectedProgress = log.getDifference(logOrder.lastKey(), new Offset(log.getFirstSegment().getName(), 0));
+    // after resuming. hard deletes should progress. Give it some time to hard delete next range
+    waitUntilExpectedProgress(expectedProgress, 1000);
+    verifyEntriesForHardDeletes(idsToDelete, true);
+
+    // test scenario:
+    // pause hard deletes, simulate a crash. Verify the resume hard deletes new deleted blobs as expected
+    index.hardDeleter.pause();
+    assertTrue("Hard deletes should have been paused ", index.hardDeleter.isPaused());
+    verifyHardDeleterThreadState(Thread.State.WAITING, hardDeleterSleepTimeWhenCaughtUpMs + 1,
+        hardDeleterSleepTimeWhenCaughtUpMs / 10);
+    // delete two entries
+    idsToDelete.clear();
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    addPutEntries(2, PUT_RECORD_SIZE, Utils.Infinite_Time);
+    idsToDelete.add(getIdToDeleteFromIndexSegment(referenceIndex.lastKey()));
+    for (MockId id : idsToDelete) {
+      addDeleteEntry(id);
+    }
+    verifyEntriesForHardDeletes(idsToDelete, false);
+    // persist index so that new entries are persisted
+    index.persistIndex();
+
+    // crash
```

Comment:
I'm not sure this simulates a crash. Could you explain how the crash is simulated?